lib/rex/post/sql/ui/console/command_dispatcher/client.rb

```diff
@@ -0,0 +1,168 @@
+
+          def cmd_shell_help
+            print_line 'Usage: shell'
+            print_line
+            print_line 'Go into a raw SQL shell where SQL queries can be executed.'
+            print_line "To exit, type 'exit', 'quit', 'end' or 'stop'."
+            print_line
+          end
+
+          def cmd_shell(*args)
+            if help_args?(args)
+              cmd_shell_help
+              return
+            end
+
+            console = shell
+            # Pass in self so that we can call cmd_query in subsequent calls
+            console.interact_with_client(client_dispatcher: self)
+          end
+
+          def normalise_sql_result(result)
+            # Let's extract the columns and rows as that's the only thing we want?
+            # We can special-case it for each result type/class.
+            normalised_sql = { columns: [], rows: [], errors: [] }
+
+            case result
+            when ::Mysql::Result # MySQL
+              normalised_sql[:columns] = result.fields.each.map(&:name)
+              normalised_sql[:rows] = result.entries
+              # MySQL errors are handled by raising an exception when querying,
+              # meaning we don't have that in the Result object and instead we use the 'when :Hash' branch
+              normalised_sql[:errors] = []
+            when ::Msf::Db::PostgresPR::Connection::Result # PostgreSQL
+              normalised_sql[:columns] = result.fields.each.map(&:name)
+              normalised_sql[:rows] = result.rows
+              # PostgreSQL errors are handled by raising an exception when querying,
+              # meaning we don't have that in the Result object and instead we use the 'when :Hash' branch
+              normalised_sql[:errors] = []
+            when ::Hash # MSSQL
+              normalised_sql[:columns] = result[:colnames]
+              normalised_sql[:rows] = result[:rows]
+              normalised_sql[:errors] = result[:errors]
+            else
+              raise 'Unknown SQL query result type'
+            end
+
+            normalised_sql
+          end
+
+          # Take in a normalised SQL result and print it.
+          # If there are any errors, print those instead.
+          # @param [Hash {Symbol => Array, Symbol => Array, Symbol => Array}] result A hash of 'rows', 'columns' and 'errors'
+          def format_result(result)
+            if result[:errors].any?
+              return "Query has failed. Reasons: #{result[:errors].join(', ')}"
+            end
+
+            number_column = ['#']
+            columns = [number_column, result[:columns]].flatten
+            rows = result[:rows].map.each_with_index do |row, i|
+              [i, row].flatten
+            end
+
+            ::Rex::Text::Table.new(
+              'Header' => 'Response',
+              'Indent' => 4,
+              'Columns' => columns,
+              'Rows' => rows
+            )
+          end
+
+          # @return [Object]
+          def cmd_query_help(examples: [])
```

Comment:
Maybe? 🤷 

```suggestion
          # @return [String]
          def format_query_help(examples: []) # Different name, to avoid the parent class and super class having different argument params
            <<~USAGE
               Usage: query
               
               Run a raw SQL query on the target.
               
               #{examples.any? && examples.map { |example| "    query #{example}" }.join("\n")}
            USAGE # Unrelated code golf
          end
          
```